File: client/java/src/avt/AvCamera.java
package avt;

import main.Canvas;

public final class AvCamera {
   private static AvCamera instance;
   public int xCam;
   public int yCam;
   public int xTo;
   public int yTo;
   private int xLimit;
   private int yLimit;
   public long timeDelay;
   private int cmvx;
   private int cmdx;
   private int cmvy;
   private int cmdy;
   private static int distance;
   private static int w;
   public static boolean disable = false;
   public static boolean isFollow;
   public Base followPlayer;

   public static AvCamera gI() {
      if (instance == null) {
         instance = new AvCamera();
      }

      return instance;
   }

   public static void setDistance(int var0) {
      distance = var0;
   }

   public final void init(int var1) {
      if (this.followPlayer != null) {
         label54: {
            isFollow = false;
            w = LoadMap.w * AvMain.hd;
            distance = Canvas.w / 10;
            if (this.followPlayer.x * AvMain.hd > Canvas.hw) {
               if (this.followPlayer.x * AvMain.hd < LoadMap.wMap * w - Canvas.hw - w) {
                  this.xTo = this.followPlayer.x * AvMain.hd - Canvas.hw;
                  break label54;
               }

               this.xTo = LoadMap.wMap * w - Canvas.w;
               if (this.xTo >= 0) {
                  break label54;
               }
            }

            this.xTo = 0;
         }

         if (Canvas.w > LoadMap.wMap * w) {
            this.xTo = -(Canvas.w - LoadMap.wMap * w) / 2;
         }

         if (Canvas.h > LoadMap.Hmap * w && (var1 - 1 == 57 || var1 - 1 == 58 || var1 - 1 == 59 || var1 - 1 == 108)) {
            this.yTo = -(Canvas.h - LoadMap.Hmap * w) / 2;
         } else {
            this.yTo = LoadMap.Hmap * w - Canvas.h;
         }

         this.xLimit = LoadMap.wMap * w - Canvas.w;
         this.yLimit = LoadMap.Hmap * w - Canvas.h;
         this.xCam = this.xTo;
         if (this.xCam < 0) {
            this.xCam = 0;
         }

         if (this.xCam > this.xLimit) {
            this.xCam = this.xLimit;
         }

         if (this.yCam > this.yLimit) {
            this.yCam = this.yLimit;
         }

         if (this.yTo > this.yLimit) {
            this.yTo = this.yLimit;
         }
      }

   }

   public final void notTrans() {
      this.xCam = this.xTo;
      this.yCam = this.yTo;
   }

   public final void updateTo() {
      if (!disable) {
         if (this.xCam != this.xTo) {
            this.cmvx = this.xTo - this.xCam << 1;
            this.cmdx += this.cmvx;
            this.xCam += this.cmdx >> 4;
            this.cmdx &= 15;
            if (this.xCam < 0) {
               this.xCam = 0;
            }

            if (this.xCam > this.xLimit) {
               this.xCam = this.xLimit;
            }
         }
      } else {
         if (this.xCam < 0) {
            this.xCam = 0;
         }

         if (this.xCam > LoadMap.wMap * LoadMap.w * AvMain.hd - Canvas.w) {
            this.xCam = LoadMap.wMap * LoadMap.w * AvMain.hd - Canvas.w;
         }
      }

      if (this.yCam != this.yTo) {
         this.cmvy = this.yTo - this.yCam << 1;
         this.cmdy += this.cmvy;
         this.yCam += this.cmdy >> 4;
         this.cmdy &= 15;
         if (this.yCam > this.yLimit) {
            this.yCam = this.yLimit;
         }
      }

   }

   public final void setToPos(int var1, int var2) {
      this.timeDelay = 0L;
      this.xTo = var1 - Canvas.hw;
      this.yTo = var2 - Canvas.hh;
      if (this.xTo < 0) {
         this.xTo = 0;
      }

      if (this.xTo > LoadMap.wMap * w - Canvas.w) {
         this.xTo = LoadMap.wMap * w - Canvas.w;
      }

      if (this.yTo > LoadMap.Hmap * w - Canvas.h) {
         this.yTo = LoadMap.Hmap * w - Canvas.h;
      }

      this.setLimit();
   }

   public final void setPos(int var1, int var2) {
      this.xCam = this.xTo = 0;
      this.yCam = this.yTo = 0;
   }

   public final void update() {
      this.updateTo();
      if (System.currentTimeMillis() / 100L - this.timeDelay >= 20L && !isFollow) {
         int var1;
         if (this.followPlayer.direct == 0) {
            var1 = this.followPlayer.x * AvMain.hd + distance;
         } else {
            var1 = this.followPlayer.x * AvMain.hd - distance;
         }

         this.xTo = var1 - Canvas.hw;
         this.yTo = (this.followPlayer.y + this.followPlayer.direct_) * AvMain.hd - (Canvas.h - (Canvas.hh - w));
         if (this.followPlayer.direct == Base.LEFT) {
            if (this.followPlayer.x * AvMain.hd < Canvas.hw) {
               this.xTo = 0;
            }
         } else if (this.followPlayer.x * AvMain.hd > LoadMap.wMap * w - Canvas.hw) {
            this.xTo = LoadMap.wMap * w - Canvas.w;
         }

         this.setLimit();
      }

   }

   private void setLimit() {
      if (LoadMap.TYPEMAP >= 0 && LoadMap.TYPEMAP < LoadMap.bg.length && LoadMap.bg[LoadMap.TYPEMAP] == -1 && LoadMap.imgBG == null && Canvas.h > LoadMap.Hmap * w) {
         this.yCam = this.yTo = -(Canvas.h - LoadMap.Hmap * w) / 2;
      }

      if (Canvas.w > LoadMap.wMap * w) {
         this.xCam = this.xTo = -(Canvas.w - LoadMap.wMap * w) / 2;
      }

   }
}

File: client/unity/Assets/Scripts/Assembly-CSharp/AvCamera.cs
using System;
using UnityEngine;

public class AvCamera
{
	public static AvCamera instance;

	public float xCam;

	public float yCam;

	public float xTo;

	public float yTo;

	public float xLimit;

	public float yLimit;

	public long timeDelay;

	private float cmvx;

	private float cmdx;

	private float cmvy;

	private float cmdy;

	public static int distance;

	public static int w;

	public static bool disable;

	public static bool isFollow;

	public static bool isMove;

	public Base followPlayer;

	public float hdFarm = 1f;

	public float vY;

	public float vX;

	public static AvCamera gI()
	{
		if (instance == null)
		{
			instance = new AvCamera();
		}
		return instance;
	}

	public static void setDistance(int dis)
	{
		distance = dis;
	}

	public void init(int index)
	{
		if (followPlayer == null)
		{
			return;
		}
		hdFarm = 1f;
		isFollow = false;
		hdFarm = AvMain.zoom;
		w = (int)((float)(LoadMap.w * AvMain.hd) * hdFarm);
		distance = Canvas.w / 20;
		if (distance < 20)
		{
			distance = 20;
		}
		if ((float)(followPlayer.x * AvMain.hd) * hdFarm > (float)Canvas.hw)
		{
			if ((float)(followPlayer.x * AvMain.hd) * hdFarm < (float)(LoadMap.wMap * w - Canvas.hw - w))
			{
				xTo = (int)((float)(followPlayer.x * AvMain.hd) * hdFarm - (float)Canvas.hw);
			}
			else
			{
				xTo = LoadMap.wMap * w - Canvas.w;
				if (xTo < 0f)
				{
					xTo = 0f;
				}
			}
		}
		else
		{
			xTo = 0f;
		}
		int hCan = Canvas.hCan;
		if (Canvas.w > LoadMap.wMap * w)
		{
			xTo = -(Canvas.w - LoadMap.wMap * w) / 2;
		}
		if (hCan > LoadMap.Hmap * w && (index - 1 == 57 || index - 1 == 58 || index - 1 == 59 || index - 1 == 108))
		{
			yTo = -(hCan - LoadMap.Hmap * w) / 2;
		}
		else
		{
			yTo = LoadMap.Hmap * w - hCan;
		}
		xLimit = (int)((float)(LoadMap.wMap * w - Canvas.w) / hdFarm);
		yLimit = (int)((float)(LoadMap.Hmap * w - hCan) / hdFarm);
		xCam = xTo;
		if (yTo > yLimit)
		{
			yTo = yLimit;
		}
		setPosFollowPlayer();
		xCam = xTo;
		yCam = yTo;
		if (xCam < 0f)
		{
			xCam = 0f;
		}
		if (xCam > xLimit)
		{
			xCam = xLimit;
		}
		if (yCam > yLimit)
		{
			yCam = yLimit;
		}
	}

	public void notTrans()
	{
		xCam = xTo;
		yCam = yTo;
	}

	public void updateTo()
	{
		float zoom = AvMain.zoom;
		if (!disable)
		{
			if (xCam != xTo)
			{
				cmvx = (int)(xTo - xCam) << 1;
				cmdx += cmvx;
				xCam += (int)cmdx >> 4;
				cmdx = (int)cmdx & 0xF;
				if (xCam < 0f)
				{
					xCam = 0f;
				}
				if (xCam > xLimit)
				{
					xCam = xLimit;
				}
			}
		}
		else
		{
			if (xCam < 0f)
			{
				xCam = 0f;
			}
			if (xCam > (float)(LoadMap.wMap * LoadMap.w * AvMain.hd) * zoom - (float)Canvas.w)
			{
				xCam = (int)((float)(LoadMap.wMap * LoadMap.w * AvMain.hd) * zoom - (float)Canvas.w);
			}
		}
		if (yCam != yTo)
		{
			cmvy = (int)(yTo - yCam) << 1;
			cmdy += cmvy;
			yCam += (int)cmdy >> 4;
			cmdy = (int)cmdy & 0xF;
			if (yCam > yLimit)
			{
				yCam = yLimit;
			}
		}
	}

	public void setToPos(float x, float y)
	{
		float zoom = AvMain.zoom;
		timeDelay = 0L;
		xTo = x - (float)Canvas.hw;
		yTo = y - (float)(Canvas.hCan / 2);
		if (xTo < 0f)
		{
			xTo = 0f;
		}
		if (xTo > (float)(LoadMap.wMap * w - Canvas.w))
		{
			xTo = LoadMap.wMap * w - Canvas.w;
		}
		if (yTo < 0f)
		{
			yTo = 0f;
		}
		int num = Canvas.hCan;
		if (TouchScreenKeyboard.visible)
		{
			num = Canvas.h - MyScreen.hTab - 17 * AvMain.hd;
		}
		if (yTo > (float)(LoadMap.Hmap * w - num))
		{
			yTo = LoadMap.Hmap * w - num;
		}
		setLimit();
	}

	public void setToPos(float x, float y, int ih)
	{
		timeDelay = 0L;
		xTo = x - (float)Canvas.hw;
		yTo = y - (float)(Canvas.hCan / 2);
		if (xTo < 0f)
		{
			xTo = 0f;
		}
		if (xTo > (float)(LoadMap.wMap * w - Canvas.w))
		{
			xTo = LoadMap.wMap * w - Canvas.w;
		}
		if (yTo < 0f)
		{
			yTo = 0f;
		}
		int num = ih - MyScreen.hTab - 17 * AvMain.hd;
		if (yTo > (float)(LoadMap.Hmap * w - num))
		{
			yTo = LoadMap.Hmap * w - num;
		}
		setLimit();
	}

	public void setPos(int x, int y)
	{
		xCam = (xTo = x);
		yCam = (yTo = y);
	}

	public void update()
	{
		if (Canvas.isZoom)
		{
			return;
		}
		if (Canvas.currentMyScreen != RaceScr.me && Canvas.welcome == null)
		{
			isFollow = true;
		}
		if (followPlayer != null && MapScr.isWedding)
		{
			isFollow = false;
		}
		if (isMove && CRes.abs((int)vX) <= 1 && CRes.abs((int)vY) <= 1)
		{
			isMove = false;
			timeDelay = Environment.TickCount / 100;
			vX = (vY = 0f);
		}
		moveCamera();
		if (vX != 0f || vY != 0f)
		{
			return;
		}
		updateTo();
		if (!isFollow)
		{
			int num = 0;
			num = ((followPlayer.direct != Base.RIGHT) ? (followPlayer.x * AvMain.hd - distance) : (followPlayer.x * AvMain.hd + distance));
			xTo = (int)((float)num - (float)Canvas.hw / hdFarm);
			int hCan = Canvas.hCan;
			yTo = (int)((float)(followPlayer.y * AvMain.hd) - ((float)hCan / hdFarm - ((float)(hCan / 2) / hdFarm - (float)(w / 3))));
			if (followPlayer.direct == Base.LEFT)
			{
				if ((float)followPlayer.x < (float)Canvas.hw / ((float)AvMain.hd * hdFarm))
				{
					xTo = 0f;
				}
			}
			else if (xTo > xLimit)
			{
				xTo = xLimit;
			}
		}
		setLimit();
	}

	public void setPosFollowPlayer()
	{
		int num = 0;
		num = ((followPlayer.direct != Base.RIGHT) ? (followPlayer.x * AvMain.hd - distance) : (followPlayer.x * AvMain.hd + distance));
		xTo = (int)((float)num - (float)Canvas.hw / hdFarm);
		int hCan = Canvas.hCan;
		yTo = (int)((float)(followPlayer.y * AvMain.hd) - ((float)hCan / hdFarm - ((float)(hCan / 2) / hdFarm - (float)(w / 3))));
		if (followPlayer.direct == Base.LEFT)
		{
			if ((float)followPlayer.x < (float)Canvas.hw / ((float)AvMain.hd * hdFarm))
			{
				xTo = 0f;
			}
		}
		else if (xTo > xLimit)
		{
			xTo = xLimit;
		}
	}

	public void setLimit()
	{
		int hCan = Canvas.hCan;
		if (LoadMap.TYPEMAP >= 0 && LoadMap.TYPEMAP < LoadMap.bg.Length && LoadMap.bg[LoadMap.TYPEMAP] == -1 && LoadMap.imgBG == null)
		{
			if (hCan > LoadMap.Hmap * w)
			{
				yCam = (yTo = -(hCan - LoadMap.Hmap * w) / 2);
			}
		}
		else if (yCam > yLimit)
		{
			yCam = (yTo = yLimit);
		}
		if (Canvas.w > LoadMap.wMap * w)
		{
			xCam = (xTo = -(Canvas.w - LoadMap.wMap * w) / 2);
		}
		else if (xCam < 0f)
		{
			xCam = (xTo = 0f);
		}
	}

	public void moveCamera()
	{
		if (Canvas.menuMain != null || Canvas.currentDialog != null)
		{
			return;
		}
		if (vY != 0f)
		{
			float num = -Canvas.hCan;
			if (yCam + vY / 15f < num + 100f)
			{
				if (yCam + vY / 15f < num)
				{
					yCam = (yTo = -Canvas.hCan);
					vY /= 6f;
					vY *= -1f;
				}
				else
				{
					vY -= vY / 20f;
				}
			}
			if (yCam + vY / 15f > yLimit - (float)(100 * AvMain.hd))
			{
				if (yCam + vY / 15f >= yLimit)
				{
					yCam = (yTo = yLimit);
					vY /= 6f;
					vY *= -1f;
				}
				else
				{
					vY -= vY / 20f;
				}
			}
			yCam += vY / 15f;
			yTo = yCam;
			vY -= vY / 20f;
		}
		if (vX == 0f)
		{
			return;
		}
		if (xCam + vX / 15f < (float)(100 * AvMain.hd))
		{
			if (xCam + vX / 15f < 0f)
			{
				xCam = (xTo = 0f);
				vX /= 6f;
				vX *= -1f;
			}
			else
			{
				vX -= vX / 20f;
			}
		}
		if (xCam + vX / 15f > xLimit - (float)(100 * AvMain.hd))
		{
			if (xCam + vX / 15f >= xLimit)
			{
				xCam = (xTo = xLimit);
				vX /= 6f;
				vX *= -1f;
			}
			else
			{
				vX -= vX / 20f;
			}
		}
		xCam += vX / 15f;
		xTo = xCam;
		vX -= vX / 20f;
	}
}
